crime-commons-spring-boot-starter-rest-client/src/main/java/uk/gov/justice/laa/crime/commons/filters/WebClientFilters.java

```diff
@@ -86,29 +88,33 @@ public static ExchangeFilterFunction handleErrorResponse() {
                 HttpStatus.SERVICE_UNAVAILABLE, HttpStatus.GATEWAY_TIMEOUT
         );
 
-        return ExchangeFilterFunctions.statusError(
-                HttpStatusCode::isError, r -> {
+        return ExchangeFilterFunction.ofResponseProcessor(
+                response -> {
+                    HttpStatus status = HttpStatus.valueOf(response.statusCode().value());
 
-                    HttpStatus status = HttpStatus.valueOf(r.statusCode().value());
-
-                    String errorMessage =
-                            String.format("Received error %s due to %s",
-                                    r.statusCode().value(), status.getReasonPhrase());
+                    String errorMessage = String.format("Received error %s due to %s",
+                            response.statusCode().value(), status.getReasonPhrase());
 
                     if (retryableStatuses.contains(status)) {
-                        return new RetryableWebClientResponseException(errorMessage);
+                        return Mono.error(new RetryableWebClientResponseException(errorMessage));
                     }
 
-                    if (status.is4xxClientError() && !status.equals(HttpStatus.NOT_FOUND)) {
-                        return new HttpClientErrorException(r.statusCode(), errorMessage);
+                    if (status.is4xxClientError()) {
+                        if (status.equals(HttpStatus.NOT_FOUND)) {
+                            return Mono.error(WebClientResponseException.create(response.statusCode().value(), status.getReasonPhrase(), null, null, null));
+                        }
+                        return Mono.error(new HttpClientErrorException(response.statusCode(), errorMessage));
                     }
 
                     if (status.is5xxServerError()) {
-                        return new HttpServerErrorException(r.statusCode(), errorMessage);
+                        Optional<Mono<ErrorDTO>> errorDTOMono = Optional.ofNullable(response.bodyToMono(ErrorDTO.class));
+                        if (HttpStatus.INTERNAL_SERVER_ERROR == response.statusCode() && errorDTOMono.isPresent()) {
+                            return errorDTOMono.get()
+                                    .flatMap(errorBody -> Mono.error(new MAATApplicationException(errorBody.getMessage())));
```

Comment:
Wont this filter apply to calls made between microservices aswell, for example c3 -> hardship in which case is MaatApplicationException the error we should be creating here or should it be more generic such as HTTPClientErrorException?

Reply:
As of now we are only dealing with the MAATApplicationException returned from the Court Data API, to limit the changes.

Reply:
Oh ok so we can expand on this later, that makes sense thanks 👍